### ISSUE TITLE
Combining spectra

### ISSUE THREAD

**Issue by jotsetung** [lgatto/MSnbase#366]
@lgatto, is there a method available in `MSnbase` to combine/merge spectra into a single one? I know that `combineSpectra` would work, but is there already some other functionality implemented that would allow to do that?

So far `combineSpectra` is tuned to be used mainly for combining profile-mode spectra, I would need to make some changes to make it applicable more generally.

**Comment by jotsetung** [comment]
I've changed `combineSpectra` supporting to merge spectra by returning the *union* of their peaks (setting the new parameter `onlyMain = FALSE`).

**Comment by lgatto** [comment]
I would also like to draw your attention to this [`combineSpectra` branch](https://github.com/lgatto/MSnbase/tree/combineSpectra) and [here](https://github.com/lgatto/MSnbase/issues/339) and [here](https://github.com/sneumann/xcms/issues/237#issuecomment-383007195).

As far as I remember, I thought that *combining spectra* was better than *merging spectra* - I would like to get your inputs @jotsetung @sgibb 

I also think that the syntax documented [above](https://github.com/sneumann/xcms/issues/237#issuecomment-383007195) and below is important.
```r
> data(itraqdata) ## MSnExp
> length(itraqdata)
[1] 55
> ## how spectra would be combined
> fData(itraqdata)$combine <- as.integer(seq(1, 10, length.out = 55))
> fData(itraqdata)$combine
 [1]  1  1  1  1  1  1  2  2  2  2  2  2  3  3  3  3  3  3  4  4  4  4  4  4  5
[26]  5  5  5  5  5  6  6  6  6  6  6  7  7  7  7  7  7  8  8  8  8  8  8  9  9
[51]  9  9  9  9 10
> x <- combineSpectra(itraqdata, mzFun = base::mean, fcol = "combine") ## new 
> ## length(x) would be 10
```
I'll follow up in #367

**Comment by lgatto** [comment]
Looking at #367, it is an implementation for a very specific use case. I think we should think on a better interface, that is general and allows the flexibility to accommodate different use situations. What about something like this, still keeping in mind that we would have a `comineSpectra` or `mergeSpectra` method, that works on an `MSnExp` (`x` below) and uses a feature variable to define the groups of spectra to be merged (`g` below). 
```r
combineSpectra(x, fcol = "g", ## as above
               fun = myCombineFun, 
               myCobineArg1 = 1,  myCobineArg2 = base::mean)
```
A `myCombineFun` would take an `Spectra` object, or the full `MSnExp`, depending, do its business, and return a new `MSnExp` that would also be returned by `combineSpectra`.

In terms of documentation, the general API would be documented in `combineSpectra`, and we could describe the use cases and arguments in great length in individual `myCombineFun` manual pages.

**Comment by lgatto** [comment]
Also, I would like to discuss the in memory vs on disk models in the frame of this particular method. Once we start to modify spectra, we won't be able to work on disk any more. I currently generate new spectra on the fly, upon modification and merging of the original data that is read from disk, but this won't scale. This re-opens the discussion of a read-write on disk backend (see for example [here](https://github.com/sneumann/mzR/issues/131)], or this could simply be serialising spectra to disk).

**Comment by jotsetung** [comment]
Trying to follow up here. Thanks for linking the various information together @lgatto I knew there was something.

As far as I understand the `combineSpectra` you propose above is something that allows to apply any function to a **group** of spectra, identified by `fcol`. I'd rather call that `applySpectraGroup`, `aggregateSpectra` or similar (bad names, I know).
I also don't yet see the benefits of this approach over `combineSpectra(<list of spectra>, ...` - one could simply split the `MSnExp` by `fcol` and `lapply` on that. 

I think we should first list the use cases we have in mind.

From my side:
1) for smoothing LC signal along retention time: average the signal (m/z and intensity) from consecutive scans/spectra in an `MSnExp`. That's currently done with `combineSpectraMovingWindow` (which uses `combineSpectra` on that).
2) as in https://github.com/sneumann/xcms/issues/321, given a list of MS2 spectra that *should* represent signal for the same ion, combine these into a single spectrum.

**Comment by lgatto** [comment]
Re naming, I don't really mind if it's *combine*, *merge* or *aggregate* (we will vote), but I **do want** a tidy (public) interface that takes and `MSnExp` as input and returns a new `MSnExp` as output, where a set of spectra have been combined based on a grouping defined as a feature variable. It is then easy to perform it on a small set of spectra to be combine by subsetting the whole experiment, and to apply it at the whole experiment level. And from there, one can easily extract an (annotated) list if needed.

My point is that I **refuse** to have to do all the plumbing myself (or make users for that matter): extract spectra, iterate/lapply, combine, and create a new experiment, .... The `applySomething` that you describe above and `lapply` suggest exactly that kind of behaviour. 

3. Another use case is [BoxCar](https://www.nature.com/articles/s41592-018-0003-5) - see Fig 1a for the type of merging that is needed.

**Comment by sgibb** [comment]
@lgatto I really like you approach. It is much easier for the user as the current implementation. Also it is inline with `combineFeatures` for `MSnSet`. So `combineSpectra` should have exact the same interface as `combineFeatures`.

**Comment by lgatto** [comment]
Yes, that's the idea. The interface won't exactly be the same
```r
> args(combineFeatures)
function (object, groupBy, fun = c("mean", "median", "weighted.mean", 
    "sum", "medpolish", "robust", "iPQF", "NTR"), fcol, redundancy.handler = c("unique", 
    "multiple"), cv = TRUE, cv.norm = "sum", verbose = isMSnbaseVerbose(), 
    ...) 
```
`groupBy` is an old one and has been replaced by `fcol`, which is used in many other functions as a way to use a feature variable. We should aim for 
```r
> args(comineSpectra)
function(object, fcol, fun, ...)
```
which is what `combineFeatures` really should be.

**Comment by jotsetung** [comment]
Wasn't aware of the `combineFeatures` method. In that sense it makes indeed sense to have `combineSpectra` doing the same. Question is what to do with the existing `combineSpectra` function? We could rename that and make it an internal function. Would also make sense to rename that function to something like `reduceSpectra`, because it is in fact reducing multiple spectra into a single one - with the same m/z range (while the BoxCar, if I understand correctly, *stitches* spectra together).

**Comment by jotsetung** [comment]
And what about defining `combineSpectra,MSnExp` and `combineSpectra,Spectra`? There are different ways to access the merging columns in `MSnExp` and `Spectra` (`fData` vs `mcol`). Also, the return object would be different (`MSnExp` *vs* `Spectra`).

The argument `fun` would then be a function that takes a `list` of spectra and returns a single spectrum, right? The name of the function is however key - I'm not so happy with `combineByMean` @sgibb, do you have eventually an idea for a name?

**Comment by lgatto** [comment]
Yes, I like `combineSpectra,MSnExp` and `combineSpectra,Spectra`. 

However, for the latter, the input is a list of spectra (object of class 'Spectra`) and the output would still be a list of spectra (of class `Spectra`), even of my initial list contains 3 items to be combined into a single spectrum. I think it is important to keep the input and output always the same. If as a user/programmer, I know I will get a single spectrum after merging, I can extract it explicitly, but at least there aren't any bad surprises when this happens unexpectedly.

**Comment by lgatto** [comment]
As for a name, the original `combineSpectra` calculates mean mz and intensity values; so what about `meanMzInts`. It's not beautiful, but it says what the function does.

**Comment by lgatto** [comment]
> The `meanMzInts` function does however take a list of spectra and combines all of them into a single spectrum (that's what I meant).

I think then it needs to be adapted to fit the new interface.

Again, what's the difference between aggregation, merging, combining, ... ? There are nuances that are field specific. What about `meanMzInts`, `sumMwInts`, `maxMzInts` for the original behaviour. There can still be a single private function that calls the different functions, but the user would get different visible functions that do exactly what is written on the tin.

**Comment by jotsetung** [comment]
> I think then it needs to be adapted to fit the new interface.

The reason is that I would need to call this function (`meanMzInts`) by the `combineSpectraMovingWindow` - and I can not use `tapply` there. That's also how the current `combineSpectra` works, it takes a `list` of spectra and returns one.

**Comment by lgatto** [comment]
Internally yes, it will take a list (corresponding to one set/factor defined by `fcol`) and return one spectrum. But these spectra (is single spectrum) will have to be returned as a list/experiment higher up. Some thing like:
```r
g <- factor(fData(object)[, fcol]
res <- vector("list", length = nlevels(g))
for (i in levels(g)) {
   x <- object[g == i]
   res[[i]] <- internalCombineFunction(x, ...)
}
```

**Comment by jotsetung** [comment]
In https://github.com/lgatto/MSnbase/commit/64b2f7350a3ddc8d9559bcc850b73787825ea96c I've implemented the meanMzInts to make space for the `combineSpectra` methods. I would then implement next the `combineSpectra,Spectra` method.

**Comment by jotsetung** [comment]
Have a look at https://github.com/lgatto/MSnbase/commit/cb4f6357ea86c21a858534ca9c8fdfddfeaa8d8f (part of PR https://github.com/lgatto/MSnbase/pull/367) where I have added `combineSpectra,Spectra`.

In this implementation I just return the `fcol` metadata column of the input `Spectra`. Returning all metadata columns doesn't seem right to me, but that is matter of discussion.

**Comment by sgibb** [comment]
:+1: for `meanMzInts`

@jotsetung in https://github.com/lgatto/MSnbase/commit/cb4f6357ea86c21a858534ca9c8fdfddfeaa8d8f you miss the `main` argument (main is essential 1 in your method).

Am I right, that we have now two different branches/implementations for the same problem?

**Comment by lgatto** [comment]
> Am I right, that we have now two different branches/implementations for the same problem?

I suggest we work on the new one, initiated by Jo, and may be keep my old branch for reference for now.

**Comment by lgatto** [comment]
> In this implementation I just return the `fcol` metadata column of the input `Spectra`. Returning all metadata columns doesn't seem right to me, but that is matter of discussion.

I would rather keep everything. It's generally easier to delete stuff than to recover it.

**Comment by jotsetung** [comment]
@sgibb re `main` argument: that argument is defined in `meanMzInts` - it can be passed in `combineSpectra` with the `...` arguments.

@lgatto re what to keep from the metadata columns. In your `combineSpectra,MSnExp` you did in fact keep all columns. My only objection to that is that that way we might keep columns with information that is relevant only to the, as in your implementation, first `Spectrum` in a set, while the resulting `Spectrum` is a combination of all spectra in the set. But again, I have no strong opinion here. If we want to keep all columns we can surely do that.

**Comment by jotsetung** [comment]
BTW, I would be more than happy if one of you implements the `combineSpectra,MSnExp` since I am not that familiar with the internals of that class...

**Comment by jotsetung** [comment]
Have a look at https://github.com/lgatto/MSnbase/commit/8a5f0720d30bf10f1b59dfd2b3bd0bb01018cce8. I addressed now both comments: `combineSpectra,Spectra` has an argument `main` that allows to specify which of the spectra in a set should be considered the *main* spectrum (@sgibb's comment). All metadata will be reported for this *main* spectrum (@lgatto's comment). Also, values from all slots except `mz` and `intensity` are taken from this *main* spectrum. I am actually pretty happy with the implementation now. Thanks to your input!

**Comment by jotsetung** [comment]
In commit https://github.com/lgatto/MSnbase/commit/78560be5bc77be8ae82c9c7afc5cd0d065d8282f I just added the option `main = "last"` to `combineSpectra,Spectra`. So, metadata can be selected for the first, middle and last spectrum from each spectrum set.

**Comment by sgibb** [comment]
@jotsetung Sry, but I am not sure that the `main` argument is useful at all. Isn't the `"first", "middle"` or `"last"` spectrum in a group more or less arbitrarily chosen? As user I can't predict the `"middle"` spectrum if I don't know the number of spectra in a group (which could be different for each group). And why should be the `"last"` one better/worse than the `"first"` one? IMHO we could drop the `main` argument and take the metadata from the first spectrum silently. If the user really want the metadata from a specific spectrum he can reconstruct it from the original `MSnExp`/`Spectra` object.

**Comment by jotsetung** [comment]
😄 I've added the `main` argument to `combineSpectra` because you asked for it in:
> @jotsetung in cb4f635 you miss the main argument (main is essential 1 in your method).

No prob, I can remove `main` again. For the `middle`: that's the default for `combineSpectraMovingWindow` as there you know what the *middle* spectrum is. I've just added `"last"` because if you have `"first"` and `"middle"` it would be only natural to have also a `"last"`. Nevermind. I remove the argument again.

**Comment by jotsetung** [comment]
Funny @sgibb , in a previous comment you mentioned that `main` is missing (*in cb4f635 you miss the main argument (main is essential 1 in your method).*) 😉 

In commit https://github.com/lgatto/MSnbase/commit/b1a04f30eda16a3de3880d5bfad6aa5cfd68883b I removed the `main` argument again. The reason for `main = "middle"` was that that was the default in `meanMzInts` (for a single list of `Spectrum` it is straight forward to know the middle spectrum - and it is also the default for `combineSpectraMovingWindow`). Having a `"first"` and `"middle"` it is just a natural consequence to have also the option `"last"`. Nevertheless, I've removed the argument from `combineSpectra` and return always the metadata (and all slots except `mz` and `intensity` always from the first spectrum in a set.

**Comment by sgibb** [comment]
@jotsetung I am sorry.

This was a hasty reaction because I saw your old method had an `main` argument and the new implementation miss it. After our discussion before I thought that `main` is important for you and want to give you the hint that I thought you missed it. But I didn't saw that `meanMzInt` has a `main` and `combineSpectra` can forward it via `...`. 

I thought first that it is really important for you to choose a specific `main` spectrum but with the introduction of `"first", "middle", "last"` it seems to me that the choice is more or less arbitrary and we could remove this argument completely. Sorry for the confusion. I didn't want to play games with you.

In conclusion: If you need the `main` argument in `meanMzInt` as `numeric` keep it there and use `...` in `combineSpectra` (as you did in your current commit).

**Comment by jotsetung** [comment]
No prob @sgibb. In the end it is just important that it is documented from which spectrum the metadata is taken. While I liked the idea that the user can choose a different spectrum from which to take the metadata you are right - in most cases *middle* or *last* is not bettern than *first*. And for implementations like `combineSpectraMovingWindow` the `main` argument is defined and passed along to `meanMzInts`.

**Comment by lgatto** [comment]
Re metadata when combining spectra, should this not be defined by the person that writes the inner functions, passed to `combineSpectra(..., fun)`? Some metadata will become wrong (for example retention time), and it's the developer that will decide whether to drop it, return one value from the original spectra, a mean value, ... and so on for all metadata. So `main` could be completely ignored. 

So my suggestion would be that if the developer has strong opinions, they hard-code what's happening with the meta-data. If not, there's a main function that, by default, takes `1`. But, as far as I understand, this `1` will be the same for all groups, right? In case of a `Spectra` or `MSnExp`, it won't be possible to set different main spectra per group. Which is fine by me (I think this would become confusing, probably for no or little gain) - just checking.

**Comment by jotsetung** [comment]
You are right, the *inner* function (passed with `fun` to `combineSpectra`) takes care of filling all slot values of the returned `Spectrum` object. A *problem* is all other metadata that is present in `mcols` of a `Spectra` (or in `featureData` of an `MSnExp`). Here we settled on returning all columns from the first element of each set. I'm OK with that solution, as trying to guess which *main* spectrum the inner function might have been used is tricky and would complicate things.

**Comment by jotsetung** [comment]
And yes, `main` is completely ignored, as it was removed from the `combineSpectra`. I still want to keep that argument in `meanMzInts` (default is `1L`) because it is key for `combineSpectraMovingWindow`.

**Comment by lgatto** [comment]
I think this can be closed, right?

**Comment by jotsetung** [comment]
Yep, from my side all OK.
